File: app/src/main/java/org/scp/gymlog/model/Exercise.java
package org.scp.gymlog.model;

import static org.scp.gymlog.util.Constants.DATE_ZERO;
import static org.scp.gymlog.util.Constants.FIVE;
import static org.scp.gymlog.util.Constants.ONE_HUNDRED;

import androidx.annotation.NonNull;

import org.scp.gymlog.room.EntityMapped;
import org.scp.gymlog.room.entities.ExerciseEntity;
import org.scp.gymlog.room.entities.ExerciseMuscleCrossRef;
import org.scp.gymlog.room.entities.SecondaryExerciseMuscleCrossRef;
import org.scp.gymlog.room.entities.VariationEntity;
import org.scp.gymlog.util.Data;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import lombok.Getter;
import lombok.Setter;

@Setter @Getter
public class Exercise implements EntityMapped<ExerciseEntity> {
	private final List<Muscle> primaryMuscles = new ArrayList<>();
	private final List<Muscle> secondaryMuscles = new ArrayList<>();
	private final List<Variation> variations = new ArrayList<>();
	private int id;
	private String name;
	private String image;
	private Calendar lastTrained = DATE_ZERO;
	private BigDecimal step = FIVE;
	private boolean requiresBar;
	private Bar bar;
	private int restTime = -1;
	private WeightSpecification weightSpec = WeightSpecification.NO_BAR_WEIGHT;

	@Override
	public ExerciseEntity toEntity() {
		ExerciseEntity entity = new ExerciseEntity();
		entity.exerciseId = id;
		entity.name = name;
		entity.image = image;
		entity.lastTrained = lastTrained;
		entity.lastStep = step.multiply(ONE_HUNDRED).intValue();
		entity.requiresBar = requiresBar;
		entity.lastWeightSpec = weightSpec;
		entity.lastRestTime = restTime;
		if (bar != null) {
			entity.lastBarId = bar.getId();
		}
		return entity;
	}

	@Override
	@SuppressWarnings("unchecked")
	public Exercise fromEntity(@NonNull ExerciseEntity entity) {
		id = entity.exerciseId;
		name = entity.name;
		image = entity.image;
		lastTrained = entity.lastTrained;
		step = BigDecimal.valueOf(entity.lastStep).divide(ONE_HUNDRED);
		requiresBar = entity.requiresBar;
		weightSpec = entity.lastWeightSpec;
		restTime = entity.lastRestTime;
		if (entity.lastBarId != null) {
			bar = Data.getBar(entity.lastBarId);
		}
		return this;
	}

	public ExerciseMuscleCrossRef[] toMuscleListEntities() {
		return primaryMuscles.stream()
				.map(Muscle::getId)
				.map(muscleId -> {
					ExerciseMuscleCrossRef xRef = new ExerciseMuscleCrossRef();
					xRef.exerciseId = id;
					xRef.muscleId = muscleId;
					return xRef;
				}).toArray(ExerciseMuscleCrossRef[]::new);
	}

	public SecondaryExerciseMuscleCrossRef[] toSecondaryMuscleListEntities() {
		return secondaryMuscles.stream()
				.map(Muscle::getId)
				.map(muscleId -> {
					SecondaryExerciseMuscleCrossRef xRef = new SecondaryExerciseMuscleCrossRef();
					xRef.exerciseId = id;
					xRef.muscleId = muscleId;
					return xRef;
				}).toArray(SecondaryExerciseMuscleCrossRef[]::new);
	}

	public VariationEntity[] toVariationListEntities() {
		return variations.stream()
				.map(v -> {
					VariationEntity variation = v.toEntity();
					variation.exerciseId = id;
					return variation;
				}).toArray(VariationEntity[]::new);
	}
}

File: app/src/main/java/org/scp/gymlog/model/Exercise.kt
package org.scp.gymlog.model

import org.scp.gymlog.exceptions.InternalException
import org.scp.gymlog.exceptions.SaveException
import org.scp.gymlog.room.EntityMappable
import org.scp.gymlog.room.entities.ExerciseEntity
import org.scp.gymlog.room.entities.ExerciseMuscleCrossRef
import org.scp.gymlog.room.entities.SecondaryExerciseMuscleCrossRef
import org.scp.gymlog.room.entities.VariationEntity
import org.scp.gymlog.util.Constants
import org.scp.gymlog.util.Data
import java.math.BigDecimal
import java.time.LocalDateTime

class Exercise() : EntityMappable<ExerciseEntity> {

	val primaryMuscles: MutableList<Muscle> = ArrayList()
	val secondaryMuscles: MutableList<Muscle> = ArrayList()
	val variations: MutableList<Variation> = ArrayList()
	var id = 0
	var name: String = ""
	var image: String = ""
	var lastTrained: LocalDateTime = Constants.DATE_ZERO

	val defaultVariation: Variation
		get() = variations.find { it.default }
			?: throw InternalException("Default variation not found for: $id")

	constructor(entity: ExerciseEntity) : this() {
		id = entity.exerciseId
		name = entity.name
		image = entity.image
		lastTrained = entity.lastTrained
	}

	override fun toEntity(): ExerciseEntity {
		if (name.isEmpty() || image.isEmpty()) {
			throw SaveException("Can't convert exercise to entity with empty name or image")
		}
		val entity = ExerciseEntity()
		entity.exerciseId = id
		entity.name = name
		entity.image = image
		entity.lastTrained = lastTrained

		defaultVariation.apply {
			entity.type = type
			entity.lastStep = step.multiply(Constants.ONE_HUNDRED).toInt()
			entity.lastWeightSpec = weightSpec
			entity.lastRestTime = restTime
			entity.lastBarId = bar?.id
		}
		return entity
	}

	fun toMuscleListEntities(): List<ExerciseMuscleCrossRef> {
		return primaryMuscles
			.map(Muscle::id)
			.map { muscleId: Int ->
				val xRef = ExerciseMuscleCrossRef()
				xRef.exerciseId = id
				xRef.muscleId = muscleId
				xRef
			}
	}

	fun toSecondaryMuscleListEntities(): List<SecondaryExerciseMuscleCrossRef> {
		return secondaryMuscles
			.map(Muscle::id)
			.map { muscleId: Int ->
				val xRef = SecondaryExerciseMuscleCrossRef()
				xRef.exerciseId = id
				xRef.muscleId = muscleId
				xRef
			}
	}

	fun toVariationListEntities(): List<VariationEntity> {
		return variations
			.map { it.toEntity().apply { exerciseId = this@Exercise.id } }
	}
}